File: packages/python-sdk/cavendo/deliverables.py
"""
Deliverables API for the Cavendo SDK.

This module provides methods for submitting and managing deliverables.
"""

from typing import TYPE_CHECKING, Any, List, Optional, Union

from .types import ActionItem, ContentType, Deliverable, DeliverableStatus, Feedback, FileAttachment

if TYPE_CHECKING:
    from .client import CavendoClient


class DeliverablesAPI:
    """
    API for managing deliverables submitted by the agent.

    This class provides methods to submit deliverables, get feedback,
    and submit revisions. All methods support both synchronous and
    asynchronous usage.
    """

    def __init__(self, client: "CavendoClient") -> None:
        """
        Initialize the Deliverables API.

        Args:
            client: The CavendoClient instance to use for API calls.
        """
        self._client = client

    def submit(
        self,
        title: str,
        task_id: Optional[int] = None,
        project_id: Optional[Union[int, str]] = None,
        content: Optional[str] = None,
        content_type: Union[str, ContentType] = ContentType.MARKDOWN,
        metadata: Optional[dict[str, Any]] = None,
        summary: Optional[str] = None,
        files: Optional[List[FileAttachment]] = None,
        actions: Optional[List[ActionItem]] = None,
        input_tokens: Optional[int] = None,
        output_tokens: Optional[int] = None,
        provider: Optional[str] = None,
        model: Optional[str] = None,
    ) -> Deliverable:
        """
        Submit a new deliverable.

        Either ``task_id`` or ``project_id`` (or both) should be provided.
        Omit ``task_id`` for standalone (project-level) deliverables.

        IMPORTANT: For HTML, JSX, CSS, JS, or any code files, use the `files`
        parameter to attach them - do NOT put code in `content`. Use `summary`
        to describe what was created (shown in Overview tab).

        Args:
            title: Title of the deliverable.
            task_id: Optional task ID (omit for standalone deliverables).
            project_id: Optional project name or ID for standalone deliverables.
            content: Optional plain text/markdown content. DO NOT use for
                HTML/code - use `files` instead.
            content_type: Type of content (markdown, html, json, text, code).
            metadata: Optional metadata (e.g., sources, references).
            summary: Text description of the work completed (markdown supported).
                This is shown in the Overview tab. For code deliverables, describe
                what was built and any key decisions made.
            files: Array of file attachments (HTML, JSX, CSS, JS, etc.).
                Each file dict should have:
                - filename: Filename with extension (e.g., "landing-page.html")
                - content: The complete file content. For binary files, prefix with "base64:".
                - mimeType: Optional MIME type. Auto-detected from extension if not provided.
            actions: Follow-up action items for the reviewer to complete.
                Each action dict should have:
                - action_text: The action item (e.g., "Review the landing page copy")
                - estimated_time_minutes: Optional estimated time in minutes (default: 25)
                - notes: Optional additional context or instructions

        Returns:
            The created Deliverable.

        Raises:
            ValidationError: If required fields are missing or invalid.
            NotFoundError: If the task doesn't exist.

        Example:
            >>> # Simple text deliverable
            >>> deliverable = client.deliverables.submit(
            ...     task_id=123,
            ...     title="Research Report",
            ...     content="## Findings\\n\\n...",
            ...     content_type="markdown",
            ...     metadata={"sources": ["https://example.com"]}
            ... )
            >>> print(f"Submitted deliverable {deliverable.id}")

            >>> # Code deliverable with files
            >>> deliverable = client.deliverables.submit(
            ...     task_id=456,
            ...     title="Landing Page",
            ...     summary="Created a responsive landing page with hero section.",
            ...     files=[
            ...         {"filename": "index.html", "content": "<html>...</html>"},
            ...         {"filename": "styles.css", "content": "body { ... }"},
            ...     ],
            ...     actions=[
            ...         {"action_text": "Review copy for brand voice", "estimated_time_minutes": 10},
            ...         {"action_text": "Test on mobile devices"},
            ...     ]
            ... )
        """
        body: dict[str, Any] = {
            "title": title,
            "contentType": (
                content_type.value if isinstance(content_type, ContentType) else content_type
            ),
        }
        if task_id is not None:
            body["taskId"] = task_id
        if project_id is not None:
            body["projectId"] = project_id
        if content is not None:
            body["content"] = content
        if metadata is not None:
            body["metadata"] = metadata
        if summary is not None:
            body["summary"] = summary
        if files is not None:
            body["files"] = files
        if actions is not None:
            body["actions"] = actions
        if input_tokens is not None:
            body["input_tokens"] = input_tokens
        if output_tokens is not None:
            body["output_tokens"] = output_tokens
        if provider is not None:
            body["provider"] = provider
        if model is not None:
            body["model"] = model

        data = self._client._request_json("POST", "/api/deliverables", json=body)
        return Deliverable.from_dict(data)

    async def submit_async(
        self,
        title: str,
        task_id: Optional[int] = None,
        project_id: Optional[Union[int, str]] = None,
        content: Optional[str] = None,
        content_type: Union[str, ContentType] = ContentType.MARKDOWN,
        metadata: Optional[dict[str, Any]] = None,
        summary: Optional[str] = None,
        files: Optional[List[FileAttachment]] = None,
        actions: Optional[List[ActionItem]] = None,
        input_tokens: Optional[int] = None,
        output_tokens: Optional[int] = None,
        provider: Optional[str] = None,
        model: Optional[str] = None,
    ) -> Deliverable:
        """
        Async version of submit().

        See submit() for documentation.
        """
        body: dict[str, Any] = {
            "title": title,
            "contentType": (
                content_type.value if isinstance(content_type, ContentType) else content_type
            ),
        }
        if task_id is not None:
            body["taskId"] = task_id
        if project_id is not None:
            body["projectId"] = project_id
        if content is not None:
            body["content"] = content
        if metadata is not None:
            body["metadata"] = metadata
        if summary is not None:
            body["summary"] = summary
        if files is not None:
            body["files"] = files
        if actions is not None:
            body["actions"] = actions
        if input_tokens is not None:
            body["input_tokens"] = input_tokens
        if output_tokens is not None:
            body["output_tokens"] = output_tokens
        if provider is not None:
            body["provider"] = provider
        if model is not None:
            body["model"] = model

        data = await self._client._request_json_async("POST", "/api/deliverables", json=body)
        return Deliverable.from_dict(data)

    def get(self, deliverable_id: int) -> Deliverable:
        """
        Get a specific deliverable by ID.

        Args:
            deliverable_id: The deliverable ID to retrieve.

        Returns:
            The requested Deliverable.

        Raises:
            NotFoundError: If the deliverable doesn't exist.

        Example:
            >>> deliverable = client.deliverables.get(456)
            >>> print(deliverable.content)
        """
        data = self._client._request_json("GET", f"/api/deliverables/{deliverable_id}")
        return Deliverable.from_dict(data)

    async def get_async(self, deliverable_id: int) -> Deliverable:
        """
        Async version of get().

        See get() for documentation.
        """
        data = await self._client._request_json_async("GET", f"/api/deliverables/{deliverable_id}")
        return Deliverable.from_dict(data)

    def get_feedback(self, deliverable_id: int) -> Optional[Feedback]:
        """
        Get feedback for a deliverable.

        Args:
            deliverable_id: The deliverable ID to get feedback for.

        Returns:
            Feedback if available, None otherwise.

        Example:
            >>> feedback = client.deliverables.get_feedback(456)
            >>> if feedback:
            ...     if feedback.status == "revision_requested":
            ...         print(f"Revision needed: {feedback.content}")
        """
        data = self._client._request_json("GET", f"/api/deliverables/{deliverable_id}/feedback")

        if not data or (isinstance(data, dict) and not data.get("feedback")):
            return None

        # API returns {id, status, feedback, reviewedBy, reviewedAt}
        return Feedback.from_dict(data)

    async def get_feedback_async(self, deliverable_id: int) -> Optional[Feedback]:
        """
        Async version of get_feedback().

        See get_feedback() for documentation.
        """
        data = await self._client._request_json_async(
            "GET", f"/api/deliverables/{deliverable_id}/feedback"
        )

        if not data or (isinstance(data, dict) and not data.get("feedback")):
            return None

        return Feedback.from_dict(data)

    def submit_revision(
        self,
        deliverable_id: int,
        content: Optional[str] = None,
        title: Optional[str] = None,
        content_type: Optional[Union[str, ContentType]] = None,
        metadata: Optional[dict[str, Any]] = None,
        summary: Optional[str] = None,
        files: Optional[List[FileAttachment]] = None,
    ) -> Deliverable:
        """
        Submit a revision for a deliverable.

        This creates a new version of the deliverable with updated content.

        IMPORTANT: For HTML, JSX, CSS, JS, or code file revisions, use the
        `files` parameter instead of `content`.

        Args:
            deliverable_id: The deliverable ID to revise.
            content: The revised content (for text/markdown). DO NOT use for
                HTML/code - use `files` instead.
            title: Optional new title (keeps original if not provided).
            content_type: Optional new content type.
            metadata: Optional updated metadata.
            summary: Text description of the revision changes (shown in Overview tab).
                For code deliverables, describe what was changed.
            files: Array of file attachments (HTML, JSX, CSS, JS, etc.).
                Each file dict should have:
                - filename: Filename with extension (e.g., "landing-page.html")
                - content: The complete file content. For binary files, prefix with "base64:".
                - mimeType: Optional MIME type. Auto-detected from extension if not provided.

        Returns:
            The updated Deliverable with incremented version.

        Raises:
            ValidationError: If neither content, summary, nor files is provided.
            NotFoundError: If the deliverable doesn't exist.

        Example:
            >>> # Get feedback and submit text revision
            >>> feedback = client.deliverables.get_feedback(456)
            >>> if feedback and feedback.status == "revision_requested":
            ...     revised = client.deliverables.submit_revision(
            ...         deliverable_id=456,
            ...         content="## Updated Findings\\n\\n..."
            ...     )
            ...     print(f"Submitted revision v{revised.version}")

            >>> # Submit code revision with updated files
            >>> revised = client.deliverables.submit_revision(
            ...     deliverable_id=789,
            ...     summary="Updated hero section per feedback.",
            ...     files=[
            ...         {"filename": "index.html", "content": "<html>...</html>"},
            ...     ]
            ... )
        """
        body: dict[str, Any] = {}
        if content is not None:
            body["content"] = content
        if title is not None:
            body["title"] = title
        if content_type is not None:
            body["contentType"] = (
                content_type.value if isinstance(content_type, ContentType) else content_type
            )
        if metadata is not None:
            body["metadata"] = metadata
        if summary is not None:
            body["summary"] = summary
        if files is not None:
            body["files"] = files

        data = self._client._request_json(
            "POST", f"/api/deliverables/{deliverable_id}/revision", json=body
        )
        return Deliverable.from_dict(data)

    async def submit_revision_async(
        self,
        deliverable_id: int,
        content: Optional[str] = None,
        title: Optional[str] = None,
        content_type: Optional[Union[str, ContentType]] = None,
        metadata: Optional[dict[str, Any]] = None,
        summary: Optional[str] = None,
        files: Optional[List[FileAttachment]] = None,
    ) -> Deliverable:
        """
        Async version of submit_revision().

        See submit_revision() for documentation.
        """
        body: dict[str, Any] = {}
        if content is not None:
            body["content"] = content
        if title is not None:
            body["title"] = title
        if content_type is not None:
            body["contentType"] = (
                content_type.value if isinstance(content_type, ContentType) else content_type
            )
        if metadata is not None:
            body["metadata"] = metadata
        if summary is not None:
            body["summary"] = summary
        if files is not None:
            body["files"] = files

        data = await self._client._request_json_async(
            "POST", f"/api/deliverables/{deliverable_id}/revision", json=body
        )
        return Deliverable.from_dict(data)

    def mine(
        self,
        status: Optional[Union[str, DeliverableStatus]] = None,
        task_id: Optional[int] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[Deliverable]:
        """
        List deliverables submitted by the current agent.

        Args:
            status: Filter by deliverable status.
            task_id: Filter by task ID.
            limit: Maximum number of deliverables to return.
            offset: Number of deliverables to skip for pagination.

        Returns:
            List of Deliverable objects.

        Example:
            >>> # Get all pending deliverables
            >>> pending = client.deliverables.mine(status="pending")

            >>> # Get deliverables needing revision
            >>> to_revise = client.deliverables.mine(status="revision_requested")
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = (
                status.value if isinstance(status, DeliverableStatus) else status
            )
        if task_id:
            params["taskId"] = task_id

        data = self._client._request_json("GET", "/api/deliverables/mine", params=params)

        deliverables_data = data if isinstance(data, list) else data.get("deliverables", [])
        return [Deliverable.from_dict(d) for d in deliverables_data]

    async def mine_async(
        self,
        status: Optional[Union[str, DeliverableStatus]] = None,
        task_id: Optional[int] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[Deliverable]:
        """
        Async version of mine().

        See mine() for documentation.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = (
                status.value if isinstance(status, DeliverableStatus) else status
            )
        if task_id:
            params["taskId"] = task_id

        data = await self._client._request_json_async(
            "GET", "/api/deliverables/mine", params=params
        )

        deliverables_data = data if isinstance(data, list) else data.get("deliverables", [])
        return [Deliverable.from_dict(d) for d in deliverables_data]

File: packages/mcp-server/src/tools/deliverables.ts
/**
 * Deliverable-related MCP tools for Cavendo Engine
 */

import { Tool } from '@modelcontextprotocol/sdk/types.js';
import { getClient, Deliverable, CavendoApiError } from '../client.js';

// ============================================================================
// Types
// ============================================================================

interface DeliverableFeedback {
  id: string;
  status: 'pending' | 'approved' | 'revision_requested' | 'rejected';
  feedback: string | null;
  reviewedBy: string | null;
  reviewedAt: string | null;
}

// ============================================================================
// Tool Definitions
// ============================================================================

export const submitDeliverableTool: Tool = {
  name: 'cavendo_submit_deliverable',
  description:
    'Submit a deliverable for a task. IMPORTANT: Use the "files" array for ANY file you want delivered with its proper ' +
    'extension — markdown (.md), HTML, CSS, JS, text, etc. The "content" field is only for plain text shown in the ' +
    'Overview tab (rendered as .txt). Use "summary" to describe what was created. After submitting, update the task status to "review".',
  inputSchema: {
    type: 'object',
    properties: {
      taskId: {
        type: 'string',
        description: 'The ID of the task this deliverable is for. Omit for standalone deliverables.',
      },
      projectId: {
        type: 'string',
        description: 'Project ID or name to associate with (if no taskId). Use for standalone deliverables.',
      },
      title: {
        type: 'string',
        description: 'A descriptive title for the deliverable (e.g., "Landing Page Mockup").',
      },
      summary: {
        type: 'string',
        description: 'Text description of the work completed (markdown supported). This is shown in the Overview tab. ' +
          'For code deliverables, describe what was built and any key decisions made.',
      },
      content: {
        type: 'string',
        description: 'Plain text shown in the Overview tab (rendered as a .txt file). For markdown, HTML, code, or any ' +
          'formatted content, use the "files" array with the appropriate extension instead.',
      },
      files: {
        type: 'array',
        description: 'Array of file attachments. Use this for ANY file you want delivered with proper rendering — ' +
          'markdown (.md), HTML, CSS, JS, text, images, etc. Each file needs filename (with extension) and content.',
        items: {
          type: 'object',
          properties: {
            filename: {
              type: 'string',
              description: 'Filename with extension (e.g., "landing-page.html", "styles.css", "App.jsx")',
            },
            content: {
              type: 'string',
              description: 'The complete file content. For binary files, prefix with "base64:".',
            },
            mimeType: {
              type: 'string',
              description: 'Optional MIME type. Auto-detected from extension if not provided.',
            },
          },
          required: ['filename', 'content'],
        },
      },
      actions: {
        type: 'array',
        description: 'Follow-up action items for the reviewer to complete.',
        items: {
          type: 'object',
          properties: {
            text: {
              type: 'string',
              description: 'The action item (e.g., "Review the landing page copy", "Test on mobile")',
            },
            estimated_time_minutes: {
              type: 'number',
              description: 'Estimated time in minutes (default: 25)',
            },
            notes: {
              type: 'string',
              description: 'Additional context or instructions',
            },
          },
          required: ['text'],
        },
      },
      metadata: {
        type: 'object',
        description: 'Optional metadata (sources, references, etc.).',
        additionalProperties: true,
      },
      inputTokens: {
        type: 'number',
        description: 'Number of input tokens used to generate this deliverable (for cost tracking).',
      },
      outputTokens: {
        type: 'number',
        description: 'Number of output tokens generated (for cost tracking).',
      },
      provider: {
        type: 'string',
        description: 'AI provider used (e.g., "anthropic", "openai").',
      },
      model: {
        type: 'string',
        description: 'Model used (e.g., "claude-opus-4", "gpt-4o").',
      },
    },
    required: ['title'],
    additionalProperties: false,
  },
};

export const getRevisionFeedbackTool: Tool = {
  name: 'cavendo_get_revision_feedback',
  description:
    'Get feedback for a deliverable that needs revision. Use this when a deliverable has status "revision_requested" ' +
    'to understand what changes are needed before submitting a revised version.',
  inputSchema: {
    type: 'object',
    properties: {
      deliverableId: {
        type: 'string',
        description: 'The ID of the deliverable to get feedback for.',
      },
    },
    required: ['deliverableId'],
    additionalProperties: false,
  },
};

export const submitRevisionTool: Tool = {
  name: 'cavendo_submit_revision',
  description:
    'Submit a revised version of a deliverable. Use this after receiving revision feedback ' +
    'to submit an updated version that addresses the requested changes. ' +
    'IMPORTANT: For HTML, JSX, CSS, JS, or code files, use the "files" array instead of "content".',
  inputSchema: {
    type: 'object',
    properties: {
      deliverableId: {
        type: 'string',
        description: 'The ID of the deliverable to revise.',
      },
      summary: {
        type: 'string',
        description: 'Text description of the revision changes (shown in Overview tab). ' +
          'For code deliverables, describe what was changed.',
      },
      content: {
        type: 'string',
        description: 'The revised content (for text/markdown). DO NOT use for HTML/code - use "files" array instead.',
      },
      files: {
        type: 'array',
        description: 'REQUIRED for code revisions. Array of file attachments (HTML, JSX, CSS, JS, etc.).',
        items: {
          type: 'object',
          properties: {
            filename: {
              type: 'string',
              description: 'Filename with extension (e.g., "landing-page.html", "styles.css")',
            },
            content: {
              type: 'string',
              description: 'The complete file content. For binary files, prefix with "base64:".',
            },
            mimeType: {
              type: 'string',
              description: 'Optional MIME type. Auto-detected from extension if not provided.',
            },
          },
          required: ['filename', 'content'],
        },
      },
      title: {
        type: 'string',
        description: 'Optional new title for the revision.',
      },
      contentType: {
        type: 'string',
        enum: ['markdown', 'html', 'json', 'text', 'code'],
        description: 'Optional new content type for the revision.',
      },
      metadata: {
        type: 'object',
        description: 'Optional additional metadata for the revision.',
        additionalProperties: true,
      },
    },
    required: ['deliverableId'],
    additionalProperties: false,
  },
};

// ============================================================================
// Tool Handlers
// ============================================================================

function formatDeliverable(deliverable: Deliverable): string {
  const parts = [
    `**${deliverable.title}**`,
    `- ID: ${deliverable.id}`,
    `- Content Type: ${deliverable.contentType}`,
    `- Status: ${deliverable.status}`,
    `- Version: ${deliverable.version}`,
    `- Task ID: ${deliverable.taskId}`,
  ];

  if (deliverable.feedback) {
    parts.push(`- Feedback: ${deliverable.feedback}`);
  }

  parts.push(`- Created: ${deliverable.createdAt}`);
  parts.push(`- Updated: ${deliverable.updatedAt}`);

  return parts.join('\n');
}

function formatFeedback(feedback: DeliverableFeedback): string {
  const parts = [
    '## Revision Feedback',
    '',
    `**Status**: ${feedback.status}`,
  ];

  if (feedback.reviewedBy) {
    parts.push(`**Reviewed By**: ${feedback.reviewedBy}`);
  }

  if (feedback.reviewedAt) {
    parts.push(`**Reviewed At**: ${feedback.reviewedAt}`);
  }

  if (feedback.feedback) {
    parts.push('');
    parts.push('### Feedback');
    parts.push(feedback.feedback);
  } else {
    parts.push('');
    parts.push('*No feedback provided.*');
  }

  return parts.join('\n');
}

export async function handleSubmitDeliverable(args: Record<string, unknown>): Promise<string> {
  const taskIdStr = args.taskId as string | undefined;
  const projectId = args.projectId as string | undefined;
  const title = args.title as string;
  const summary = args.summary as string | undefined;
  const content = args.content as string | undefined;
  const files = args.files as Array<{ filename: string; content: string; mimeType?: string }> | undefined;
  const actions = args.actions as Array<{ text: string; estimated_time_minutes?: number; notes?: string }> | undefined;
  const metadata = args.metadata as Record<string, unknown> | undefined;
  const inputTokens = args.inputTokens as number | undefined;
  const outputTokens = args.outputTokens as number | undefined;
  const provider = args.provider as string | undefined;
  const model = args.model as string | undefined;

  if (!title) {
    return 'Error: title is required.';
  }

  // Require at least summary, content, or files
  if (!summary && !content && (!files || files.length === 0)) {
    return 'Error: At least one of summary, content, or files is required.';
  }

  // Parse taskId if provided
  let taskId: number | undefined;
  if (taskIdStr) {
    taskId = parseInt(taskIdStr, 10);
    if (isNaN(taskId)) {
      return 'Error: taskId must be a valid number.';
    }
  }

  const client = getClient();

  try {
    // Map actions to expected format
    const mappedActions = actions?.map(a => ({
      action_text: a.text,
      estimated_time_minutes: a.estimated_time_minutes,
      notes: a.notes,
    }));

    const deliverable = await client.submitDeliverable({
      taskId,
      projectId,
      title,
      summary,
      content,
      files,
      actions: mappedActions,
      metadata,
      inputTokens,
      outputTokens,
      provider,
      model,
    });

    let response = `Deliverable submitted successfully:\n\n${formatDeliverable(deliverable)}`;

    if (taskId) {
      response += '\n\nRemember to update the task status to "review" when all deliverables are submitted.';
    } else {
      response += '\n\nNote: This deliverable was submitted without a linked task. A reviewer can link it to a task later.';
    }

    return response;
  } catch (error) {
    if (error instanceof CavendoApiError) {
      if (error.statusCode === 404) {
        return taskId ? `Task not found: ${taskId}` : `Project not found: ${projectId}`;
      }
      return `Error submitting deliverable: ${error.message} (HTTP ${error.statusCode})`;
    }
    return `Error: ${error instanceof Error ? error.message : 'Unknown error'}`;
  }
}

export async function handleGetRevisionFeedback(args: Record<string, unknown>): Promise<string> {
  if (!args.deliverableId) {
    return 'Error: deliverableId is required.';
  }

  const deliverableId = String(args.deliverableId);
  const client = getClient();

  try {
    const feedback = await client.getDeliverableFeedback(deliverableId);
    return formatFeedback(feedback as DeliverableFeedback);
  } catch (error) {
    if (error instanceof CavendoApiError) {
      if (error.statusCode === 404) {
        return `Deliverable not found or no feedback available: ${deliverableId}`;
      }
      return `Error getting feedback: ${error.message} (HTTP ${error.statusCode})`;
    }
    return `Error: ${error instanceof Error ? error.message : 'Unknown error'}`;
  }
}

export async function handleSubmitRevision(args: Record<string, unknown>): Promise<string> {
  if (!args.deliverableId) {
    return 'Error: deliverableId is required.';
  }

  const deliverableId = String(args.deliverableId);
  const summary = args.summary as string | undefined;
  const content = args.content as string | undefined;
  const files = args.files as Array<{ filename: string; content: string; mimeType?: string }> | undefined;
  const title = args.title as string | undefined;
  const contentType = args.contentType as 'markdown' | 'html' | 'json' | 'text' | 'code' | undefined;
  const metadata = args.metadata as Record<string, unknown> | undefined;

  // Require at least summary, content, or files
  if (!summary && !content && (!files || files.length === 0)) {
    return 'Error: At least one of summary, content, or files is required.';
  }

  const client = getClient();

  try {
    const deliverable = await client.submitRevision(deliverableId, {
      summary,
      content,
      files,
      title,
      contentType,
      metadata,
    });

    return `Revision submitted successfully:\n\n${formatDeliverable(deliverable)}\n\nThe deliverable is now at version ${deliverable.version}.`;
  } catch (error) {
    if (error instanceof CavendoApiError) {
      if (error.statusCode === 404) {
        return `Deliverable not found: ${deliverableId}`;
      }
      return `Error submitting revision: ${error.message} (HTTP ${error.statusCode})`;
    }
    return `Error: ${error instanceof Error ? error.message : 'Unknown error'}`;
  }
}

// ============================================================================
// Export
// ============================================================================

export const deliverableTools = [
  submitDeliverableTool,
  getRevisionFeedbackTool,
  submitRevisionTool,
];

export const deliverableHandlers: Record<string, (args: Record<string, unknown>) => Promise<string>> = {
  cavendo_submit_deliverable: handleSubmitDeliverable,
  cavendo_get_revision_feedback: handleGetRevisionFeedback,
  cavendo_submit_revision: handleSubmitRevision,
};
